kernel/src/util/ring_buffer.rs

```diff
@@ -0,0 +1,569 @@
+                .write_slice(segment_index, &items[..rb.cap - write_index])?;
+            rb.segment.write_slice(0, &items[rb.cap - write_index..])?;
+        } else {
+            rb.segment.write_slice(segment_index, items)?;
+        }
+
+        rb.advance_write_index(write_index, nitems);
+        Ok(())
+    }
+}
+
+impl<R: Deref<Target = RingBuffer<u8>>> Producer<u8, R> {
+    /// Writes data from the [`VmReader`] to the [`RingBuffer`].
+    ///
+    /// Returns the number of bytes written.
+    pub fn write(&mut self, reader: &mut VmReader) -> usize {
+        let rb = &self.rb;
+
+        let free_len = rb.free_len();
+        if free_len == 0 {
+            return 0;
+        }
+        let write_len = reader.remain().min(free_len);
+
+        let write_index = rb.write_index();
+        let write_len = if write_index + write_len > rb.cap {
+            // Write into two separate parts
+            let mut writer = rb
+                .segment
+                .writer()
+                .skip(write_index)
+                .limit(rb.cap - write_index);
+            let mut len = match writer.write_fallible(reader) {
+                Ok(len) => len,
+                Err((_, len)) => {
+                    rb.advance_write_index(write_index, len);
+                    return len;
+                }
+            };
+
+            let mut writer = rb
+                .segment
+                .writer()
+                .limit(write_len - (rb.cap - write_index));
+            len += match writer.write_fallible(reader) {
+                Ok(len) => len,
+                Err((_, len)) => len,
+            };
+            len
+        } else {
+            let mut writer = rb.segment.writer().skip(write_index).limit(write_len);
+            match writer.write_fallible(reader) {
+                Ok(len) => len,
+                Err((_, len)) => len,
+            }
+        };
+
+        rb.advance_write_index(write_index, write_len);
+        write_len
+    }
+}
+
+#[inherit_methods(from = "self.rb")]
+impl<T, R: Deref<Target = RingBuffer<T>>> Producer<T, R> {
+    pub fn capacity(&self) -> usize;
+    pub fn is_empty(&self) -> bool;
+    pub fn is_full(&self) -> bool;
+    pub fn len(&self) -> usize;
+    pub fn free_len(&self) -> usize;
+}
+
+impl<T: Pod, R: Deref<Target = RingBuffer<T>>> Consumer<T, R> {
+    const T_SIZE: usize = core::mem::size_of::<T>();
+
+    /// Pops an item from the [`RingBuffer`].
+    ///
+    /// Returns an error if the ring buffer is empty or
+    /// the underlying memory read failed.
+    pub fn pop(&mut self) -> Result<T> {
+        let rb = &self.rb;
+        if rb.is_empty() {
+            return_errno_with_message!(Errno::ENODATA, "ring buffer is empty");
+        }
+
+        let read_index = rb.read_index();
+        debug_assert!(read_index < rb.cap);
+
+        let segment_index = read_index * Self::T_SIZE;
+        let mut reader = rb.segment.reader().skip(segment_index).limit(Self::T_SIZE);
```

Comment:
All calls to `limit` in this module is basically useless. Because the module itself has ensured that the readers/writers will not be used to read/write beyond the intended range.

Reply:
The `limit` seems unavoidable in `read_fallible`/`write_fallible` because 1. the real capacity of the underlying `segment` may not be equal to the user-given capacity; 2. the actual read/write length not always equals to the parametered reader/writer's length, it also checks current ring buffer's available length.